lib/accesscontrol/records.rb

```diff
@@ -0,0 +1,55 @@
+module AccessControl
+
+  # Permission checks directly related to specific ActiveRecord records happen here.
+  module Records
+    module_function
+
+    # Ask whether the actor has permission to perform action_id
+    # on a given record.
+    #
+    # @param actor [ActiveRecord::Base] The actor we're checking for permission on.
+    # @param action_id [Integer, Array<Integer>] The action or actions we're checking whether the actor has. If this is an array, then the check is ORed.
+    # @param object_type [ActiveRecord::Base] The ActiveRecord model which we're checking for permission on.
+    # @param object_id [Integer] The id of the ActiveRecord object which we're checking for permission on.
+    # @return [Boolean] Returns true if actor has been granted the permission on the specified record, false otherwise.
+    #
+    # @example
+    #   # Can the user perform the action with id 5 for the Post with id 7?
+    #   AccessControl::Records.can?(user, 5, Post, 7)
+    def can?(actor, action_id, object_type, object_id)
+      PermittedActionOnObject.where(
```

Comment:
Kind of the same thought around these... maybe instead of wrapping this query here in this module it's actually a part of `PermittedActionOnObject` like `PermittedActionOnObject.any_permissions_on?(actor, action_id, object_type, object_id)`

Here is a more radical idea!!! I'm thinking out loud for a moment...  what if it would be better to just expose those different ways of retrieving the permissions as a part of our public api? It has been attractive to have just a simple `.can?` at the top level... but from some of our discussions it's already becoming apparent that maybe we would want some more flexibility in how that works. Is it that the user as any of the given permissions or all of them? If we expose different ways to query that at as a part of the public api, then the consumer of the api can choose. Instead of `AccessControl.can?` they could use AccessControl::PermittedActionOnObject.any_permissions_on?` or maybe we have an object representing the query process like `AccessControl::RecordPermission.for(user, Post, 7).can?(5)` or `AccessControl::RecordPermission.new(user, Post, 7).can_all_of?([5,7,9])` This also eliminates having to have logic inside can that keys off of whether we are supplying an object or not (that if statement that we were talking about yesterday). 

Quite a bit different from where we initially were... but just trying to think outside the box while we still can :-) Love to hear y'alls thoughts!

Reply:
The way we're implementing things does provide calling things on the `PermittedAction` and `PermittedActionOnObject` models directly. But I'd prefer to hide the internals as much as possible to keep what developers have to know to a minimum.

So I think that means we provide the simple interface but leave it open for power developers to go crazy with the permission models directly.

As for putting some methods on the models... I'm on the fence. I could see it work.

Reply:
I see what you are saying... it would be nice to keep the interface as clean as possible 🤔 

Another nice thing about moving the parts where we query active record to the models is it keep a little separation between the model and it knowing how to look up records the other parts of the system. In addition to keeping the query logic closer to where it feels like it belongs (on the model) it would abstract those hard to isolate finder methods to make testing easier (then you could, for example, stub `any_permissions_on?` instead of having to stub `where` and `exists?` and whatever else)?